9 statistics view

amadulhaxxani opened the pull request:
- added a new statistics page for views and downloads for items
- user can click the statistics button to view the page
- clicking a specific year or month should zoom in 
- created subscribe and unsubscribe button
- added tests

amadulhaxxani committed:
adding buttons for navigation to statistics page

adding buttons for navigation to statistics page

amadulhaxxani committed:
changing the path to views-downloads-statistics

changing the path to views-downloads-statistics

amadulhaxxani committed:
adding the new component for navigation

adding the new component for navigation to statistics

amadulhaxxani committed:
calling the ep to fetch data

calling the ep to fetch data

amadulhaxxani committed:
adding mock data created by AI for testing

adding mock data created by AI for testing

amadulhaxxani committed:
showing the graph for views and downloads

showing the graph for views and downloads

amadulhaxxani committed:
adding and fixing some design to make it look like apps part

adding and fixing some design to make it look like apps part

amadulhaxxani committed:
commenting some code

commenting some code

amadulhaxxani committed:
adding the model and resource type

adding the model and resource type

amadulhaxxani committed:
adding service for sub/unsub to item

adding service for subscribing or unsubscribing to item

amadulhaxxani committed:
adding views downloads component to show chart

adding views downloads component to show chart

amadulhaxxani committed:
adding subscribe unsubscribe buttons

adding subscribe unsubscribe buttons
and showing only for authenticated users

amadulhaxxani committed:
added translations

added translations

amadulhaxxani committed:
adding translations and removing extra comments

adding translations and removing extra comments

amadulhaxxani committed:
adding translations

adding translations

amadulhaxxani committed:
removed mock data and the implementation

removed mock data and the implementation

amadulhaxxani committed:
adding tests for sub/unsub button and charts view

adding tests for sub/unsub button and charts view

amadulhaxxani committed:
Merge branch 'clarin-v7' into 9-statistics-view

amadulhaxxani committed:
fixed copilot suggestions

fixed copilot suggestions

amadulhaxxani committed:
Merge branch '9-statistics-view' of https://github.com/ufal/dspace-angular into 9-statistics-view

amadulhaxxani commented on package.json:
```diff
@@ -52,7 +52,8 @@
     "webdriver-manager": "^12.1.8",
-    "ts-node": "10.2.1"
+    "ts-node": "10.2.1",
+    "@types/d3-dispatch": "3.0.6"
```
fixed

amadulhaxxani commented on src/app/item-page/views-downloads-statistics/views-downloads-statistics.service.ts:
```diff
@@ -0,0 +1,176 @@
+})
+export class ViewsDownloadsStatisticsService {
+  private USE_MOCK_DATA = true;
+
```
fixed

amadulhaxxani commented on src/config/app-config.interface.ts:
```diff
@@ -51,6 +52,7 @@ interface AppConfig extends Config {
   comcolSelectionSort: DiscoverySortConfig;
   signpostingEnabled: boolean;
   matomo: MatomoConfig;
+  statistics: StatisticsConfig
```
fixed

amadulhaxxani commented on src/app/item-page/simple/item-types/publication/publication.component.html:
```diff
@@ -107,10 +107,14 @@
       <a [routerLink]="[itemPageRoute + '/full']">
         {{"item.page.link.full" | translate}}
       </a>
+      <!-- <button class="btn btn-outline-info btn-sm">Subscribe</button> -->
```
removed

amadulhaxxani committed:
removed the comma

removed the comma

amadulhaxxani committed:
added translation for the buttons

added translation for the buttons

amadulhaxxani committed:
lint fix

lint fix

amadulhaxxani committed:
lint fixes

lint fixes

amadulhaxxani committed:
removing extra line of codes

removing extra line of codes

amadulhaxxani committed:
adding back the accidentally removed code

adding back the accidentally removed code

amadulhaxxani committed:
fixing the translate pipe mocking issue

fixing the translate pipe mocking issue

amadulhaxxani committed:
remove the pipe and unnecessary tests

remove the pipe and unnecessary tests

amadulhaxxani commented on src/app/item-page/full/full-item-page.component.html:
```diff
@@ -32,11 +32,19 @@
+          </div>
+          <div class="d-flex flex-row gap-1 align-items-end" *ngIf="isAuthenticated$ | async">
+              <ds-matomo-subscription-button [item]="item"></ds-matomo-subscription-button>
+              <button class="btn btn-secondary" [routerLink]="['/items',item.id, 'views-downloads-statistics']"><i class="fa fa-bar-chart"></i> View statistics</button>
```
fixed

kosarko committed:
Merge branch 'clarin-v7' into 9-statistics-view

kosarko commented on src/environments/environment.production.ts:
```diff
@@ -10,4 +10,9 @@ export const environment: Partial<BuildConfig> = {
   },
+
+  statistics: {
+    baseUrl: 'https://lindat.mff.cuni.cz/statistics',
```
This should be something to be configured per instance in `config.prod.yml` or some other way.

amadulhaxxani replied:
@kosarko 
### Configuration for Third-Party Deployments

**For LINDAT instances:** No configuration needed - the production config already includes the statistics service URL.

**For other instances:** To use this feature, configuration might be required for the statistics service endpoint:

#### Configuration File (Not exactly sure about it but it can be added)

Add to `config/config.prod.yml` or `config/config.yml`:

```yaml
# Statistics configuration
statistics:
  baseUrl: 'https://your-statistics-service.example.com/statistics'
  endpoint: '/handle'
``` 
NOTE: I am still not sure as I haven't looked deeply into the configuration files but this will probably require changes in order to make the added configuations work: 
```yaml
  let url =  `statistics/handle?h=${handle}`;  TO let url = `${appConfig.statistics.baseUrl}${appConfig.statisctics.endpoint}?h=${handle}`
 ```

kosarko commented on src/app/item-page/simple/item-types/publication/publication.component.html:
```diff
@@ -107,10 +107,13 @@
         {{"item.page.link.full" | translate}}
       </a>
+      <button class="btn btn-secondary btn-sm ml-auto" [routerLink]="['/items', object.id, 'views-downloads-statistics']">
+        <i class="fa fa-bar-chart"></i> {{ 'item.page.link.statistics' | translate }}
```
`item.page.link.statistics` is missing from the *.json5 files; is that the correct key?

amadulhaxxani replied:
I made this change blindly as copilot's suggestion. I fixed it now

kosarko commented on src/app/entity-groups/journal-entities/item-pages/journal/journal.component.spec.ts:
Why this file needs changing?

amadulhaxxani replied:
This is one of the file which uses same structure as other item types and I wasn't sure if we need to show the buttons here or not.
so as you know where are:
- untyped-item
- full-item-page
- and then this publication as well... (I can remove it as you also said that we are mainly using the untyped and then for full view full-item component)

amadulhaxxani replied:
removed the changes from this file

kosarko reviewed (changes requested):
Looks nice
- I'd move the subscribe/unsubscribe button into the statistics view (seems more obvious you are subscribing to the stats and not item changes)
- I'd move the statistics button next to the create new version, and edit item buttons
- the tooltips on point hover are all over the place sometimes outside the viewport

amadulhaxxani committed:
moving the sub/unsub button into stats page

moving the sub/unsub button into stats page and creating separate comp for the stats buttons

amadulhaxxani committed:
removing the hardcoded values for statistics ep

removing the hardcoded values for statistics ep from config files

amadulhaxxani committed:
fixing the tooltip position in the stats view

fixing the tooltip position in the stats view

amadulhaxxani committed:
lint fixes and removing statistics config

lint fixes and removing statistics config

amadulhaxxani committed:
removing unused variables

removing unused variables

amadulhaxxani committed:
configuring the statistics without hardcoding the production ep

configuring the statistics without hardcoding the production ep

amadulhaxxani committed:
unnecessary comma removal

unnecessary comma removal

amadulhaxxani committed:
copilot fixes

copilot fixes

kuchtiak-ufal commented on src/assets/i18n/en.json5:
```diff
@@ -4550,6 +4550,62 @@
+
+  "statistics.views-downloads.all-years": "All Years",
+
+  "statistics.views-downloads.months.january": "January",
```
Isn't there a smarter way to define i18n months, e.g.
"statistics.views-downloads.months": "January, February, March, ..., December",
(then split the string to array and search by index)

amadulhaxxani replied:
it's a JSON,
I never worked so much with i18n and json like this but if you mean a nested json object, I can try to implement it

kuchtiak-ufal replied:
I meant having one i18n key for all months (where months will be comma separated), to avoid so many keys.
But don't know how this would work with JavaScript, and also with some languages, e.g. Arabian ?!

amadulhaxxani replied:
Okay maybe it's not recommended to use the translations the way you are suggesting
i18n is designed for individual translation unit, it will make it more complicated. still not exactly sure about it

kuchtiak-ufal replied:
OK, so let it as it is.

kuchtiak-ufal commented on src/assets/i18n/en.json5:
```diff
@@ -6164,4 +6220,30 @@
+
+  "matomo.subscription.subscribe.error.message": "Unable to subscribe to item statistics. Please try again later.",
+
+  "matomo.subscription.unsubscribe.success.title": "Unsubscribed!",
```
(MInor) why this is with exclamation mark: "Unsubscribed!", while the "Subscribed" is without ?

amadulhaxxani replied:
I will fix it :)

kosarko commented:
as discussed - don't show the statistics button if there's no configuration (baseUrl) for it...

kuchtiak-ufal commented on src/app/item-page/views-downloads-statistics/views-downloads-statistics.service.ts:
```diff
@@ -0,0 +1,181 @@
+    @Inject(APP_CONFIG) private appConfig: AppConfig
+  ) {
+    console.log('Statistics Config:', this.appConfig.statistics);
+    this.baseUrl = this.appConfig.statistics?.baseUrl || 'http://localhost:8080/server/api/statistics';
```
What is this localhost url good for ?

amadulhaxxani replied:
I am removing it now
it's not useful or good here

amadulhaxxani committed:
showing statistics button only when baseurl and endpoint are available

showing statistics button only when baseurl and endpoint are available

amadulhaxxani committed:
fixed translations and tests

fixed translations and tests

amadulhaxxani committed:
fixing the test files

fixing the test files

kosarko committed:
updated czech translation
